internal/expressions/function/functions.go

```diff
@@ -143,6 +144,70 @@ func ChartFrom(
 	)
 }
 
+func FreightMetadata(
+	ctx context.Context,
+	c client.Client,
+	project string,
+) expr.Option {
+	return expr.Function(
+		"freightMetadata",
+		freightMetadata(ctx, c, project),
+		new(func(freightRefName, key string) any),
+	)
+}
+
+func freightMetadata(
+	ctx context.Context,
+	c client.Client,
+	project string,
+) exprFn {
+	return func(a ...any) (any, error) {
+		if len(a) != 2 {
+			return nil, fmt.Errorf("expected 2 argument, got %d", len(a))
+		}
+
+		freightRefName, ok := a[0].(string)
+		if !ok {
+			return nil, fmt.Errorf("argument must be string, got %T", a[0])
+		}
+
+		if freightRefName == "" {
+			return nil, fmt.Errorf("freight ref name must not be empty")
+		}
+
+		key, ok := a[1].(string)
+		if !ok {
+			return nil, fmt.Errorf("argument must be string, got %T", a[1])
+		}
+		if key == "" {
+			return nil, fmt.Errorf("metadata key must not be empty")
+		}
+
+		freightData := kargoapi.Freight{}
+
+		if err := c.Get(ctx, client.ObjectKey{
+			Namespace: project,
+			Name:      freightRefName,
+		}, &freightData); err != nil {
+			if kubeerr.IsNotFound(err) {
+				return nil, fmt.Errorf("freight %s not found in project %s", freightRefName, project)
```

Comment:
It might be better to return nil instead of an error. Handling errors within an expression is hard, but handling an unexpected nil using [nil-coalescing](https://expr-lang.org/docs/language-definition#nil-coalescing) is easier.

cc @hiddeco for confirmation of this.